### PR TITLE
Feat/reduce constraints

### PR DESCRIPTION
**Pull request by porcuquine**
This reduces the number of constraints to just those for the S-boxes. Intermediate results of MDS products accumulate in linear combinations, paired with intermediate results. We allocate a variable corresponding to these well-formed linear combinations and their associated value before each non-linear S-box transformation, which enforce the non-linear constraints. A final constraint is enforced on the result value extracted from the first element of the state.

### CODE REVIEW

File: src/circuit.rs [hunk 1]
```diff
@@ -1,14 +1,163 @@
...
 use std::marker::PhantomData;
 
-#[derive(Clone)]
+/// Similar to `num::Num`, we use `Elt` to accumulate both values and linear combinations, then eventually
+/// extract into a `num::AllocatedNum`, enforcing that the linear combination corresponds to the result.
+/// In this way, all intermediate calculations are accounted for, with the restriction that we can only
+/// accumulate linear (not polynomial) constraints. The set of operations provided here ensure this invariant is maintained.
+enum Elt<E: Engine> {
+    Allocated(AllocatedNum<E>),
+    Num(E::Fr, LinearCombination<E>),
```
Comment (dignifiedquire) [inline src/circuit.rs]: this needs to be `Num(Option<E::Fr>, LinearCombination<E>)` in order to properly handle blank circuits

File: src/circuit.rs [hunk 2]
```diff
@@ -36,9 +185,15 @@ where
...
-    pub fn new(elements: Vec<AllocatedNum<E>>, constants: &'a PoseidonConstants<E, Arity>) -> Self {
+    pub fn new(
+        allocated_nums: Vec<AllocatedNum<E>>,
+        constants: &'a PoseidonConstants<E, Arity>,
+    ) -> Self {
         let width = constants.width();
-
+        let elements = allocated_nums
+            .iter()
+            .map(|a| Elt::Allocated(a.to_owned()))
```
Comment (dignifiedquire) [inline src/circuit.rs]: I think you can write this as `allocated_nums.into_iter().map(Elt::Allocated).collect()`

File: src/circuit.rs (hunk 1 above)
Comment (porcuquine) [reply src/circuit.rs]: Thanks, @dignifiedquire. This is what I was trying to figure out when I interrogated you about that. Will add.

File: src/circuit.rs (hunk 2 above)
Comment (porcuquine) [reply src/circuit.rs]: Hot damn, you're right.

File: src/circuit.rs (hunk 1 above)
Comment (porcuquine) [reply src/circuit.rs]: Fixed.

File: src/circuit.rs [hunk 3]
```diff
@@ -1,14 +1,167 @@
...
+
+    fn ensure_allocated<CS: ConstraintSystem<E>>(
+        &self,
+        cs: &mut CS,
+        enforce: bool,
+    ) -> Result<AllocatedNum<E>, SynthesisError> {
+        match self {
+            Self::Allocated(v) => Ok(v.clone()),
+            Self::Num(Some(fr), lc) => {
+                let v = AllocatedNum::alloc(cs.namespace(|| "allocate for Elt::Num"), || Ok(*fr))?;
```
Comment (dignifiedquire) [inline src/circuit.rs]: this should pass `fr.unwrap_or_else(|| SynthesisError::AssignmentMissing)` instead of matching on `Some(fr)`

Comment (porcuquine) [reply src/circuit.rs]: Why? That doesn't compile (as written above). If the `Some(fr)` doesn't match, then the error will be returned from the `_` pattern below.

**Pull request merged**